jax/_src/random.py

```diff
@@ -93,170 +56,27 @@ def _asarray(x):
-
-def _threefry2x32_lowering(key1, key2, x1, x2, use_rolled_loops=True):
-  """Apply the Threefry 2x32 hash.
-
-  Args:
-    keypair: a pair of 32bit unsigned integers used for the key.
-    count: an array of dtype uint32 used for the counts.
-
-  Returns:
-    An array of dtype uint32 with the same shape as `count`.
-  """
-  x = [x1, x2]
-
-  rotations = [np.array([13, 15, 26, 6], dtype=np.uint32),
-               np.array([17, 29, 16, 24], dtype=np.uint32)]
-  ks = [key1, key2, key1 ^ key2 ^ np.uint32(0x1BD11BDA)]
-
-  x[0] = x[0] + ks[0]
-  x[1] = x[1] + ks[1]
 
-  if use_rolled_loops:
-    x, _, _ = lax.fori_loop(0, 5, rolled_loop_step, (x, rotate_list(ks), rotations))
+### prng key operations
 
-  else:
-    for r in rotations[0]:
-      x = apply_round(x, r)
-    x[0] = x[0] + ks[1]
-    x[1] = x[1] + ks[2] + np.uint32(1)
-
-    for r in rotations[1]:
-      x = apply_round(x, r)
-    x[0] = x[0] + ks[2]
-    x[1] = x[1] + ks[0] + np.uint32(2)
-
-    for r in rotations[0]:
-      x = apply_round(x, r)
-    x[0] = x[0] + ks[0]
-    x[1] = x[1] + ks[1] + np.uint32(3)
-
-    for r in rotations[1]:
-      x = apply_round(x, r)
-    x[0] = x[0] + ks[1]
-    x[1] = x[1] + ks[2] + np.uint32(4)
-
-    for r in rotations[0]:
-      x = apply_round(x, r)
-    x[0] = x[0] + ks[2]
-    x[1] = x[1] + ks[0] + np.uint32(5)
-
-  return tuple(x)
-
-
-def _threefry2x32_gpu_translation_rule(c, k1, k2, x1, x2):
-  shape = lax.broadcast_shapes(
-      c.get_shape(k1).dimensions(), c.get_shape(k2).dimensions(),
-      c.get_shape(x1).dimensions(), c.get_shape(x2).dimensions())
-  rank = len(shape)
-  if 0 in shape:
-    zeros = xla_client.ops.Broadcast(
-        xla_bridge.constant(c, np.array(0, np.uint32)), shape)
-    return xla_client.ops.Tuple(c, [zeros, zeros])
-  def _broadcast(x):
-    ndims = c.get_shape(x).rank()
-    return xla_client.ops.BroadcastInDim(x, shape,
-                                         tuple(range(rank - ndims, rank)))
-  return cuda_prng.threefry2x32(
-      c, (_broadcast(k1), _broadcast(k2)), (_broadcast(x1), _broadcast(x2)))
-
-threefry2x32_p = core.Primitive("threefry2x32")
-threefry2x32_p.multiple_results = True
-threefry2x32_p.def_impl(partial(xla.apply_primitive, threefry2x32_p))
-threefry2x32_p.def_abstract_eval(_threefry2x32_abstract_eval)
-batching.defbroadcasting(threefry2x32_p)
-xla.translations_with_avals[threefry2x32_p] = xla.lower_fun(
-    partial(_threefry2x32_lowering, use_rolled_loops=False),
-    multiple_results=True, with_avals=True)
-xla.backend_specific_translations['cpu'][threefry2x32_p] = xla.lower_fun(
-    partial(_threefry2x32_lowering, use_rolled_loops=True),
-    multiple_results=True)
-if cuda_prng:
-  xla.backend_specific_translations['gpu'][threefry2x32_p] = \
-      _threefry2x32_gpu_translation_rule
-
-@jit
-def threefry_2x32(keypair, count):
-  """Apply the Threefry 2x32 hash.
+def PRNGKey(seed: int, prng: PRNG = default_prng) -> jnp.ndarray:
```

Comment:
I had in mind something slightly different, where `PRNGKey` would return an instance of a pytree-style class with the various methods.

In your version you have to pass two object along to every PRNG consumer: the array data and a `prng` implementation with no data. Wouldn't it be cleaner to package the two together as a class?

Reply:
Yes, we have the same plan in mind. Implementing the pytree-style class is a bit more involved and requires downstream fixes. I wasn't sure how long those would take at first, so I drafted this experiment only to test out the idea that threading an implementation (by any means) achieves what we're after. It mostly seems to work, but there was one lesson learned: handling `gamma` is trickier than the rest because it relies on a primitive `random_gamma_p`. The primitive isn't as easy to parameterize by a PRNG implementation, both because it is a primitive (defined at module-level and can only be bound with Jaxpr-friendly parameters), and because its implementation is tied to the PRNG key being a `u32[2]`.

I've since implemented the pytree class approach, and pushed it just now, with downstream fixes such that our tests pass locally. This is a good time to look again if you're curious.

The `gamma` issue is still unresolved at the moment. I raise `NotImplementedError` if the PRNG isn't the default one. Ideas welcome there.